### PR TITLE
monsters can now cast spells as special attacks

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary

```SUMMARY: Infrastructure "Monsters can cast spells as a special attack"```
#### Purpose of change

@I-am-Erk requested that i make it possible for monsters to cast spells. In doing so, i realized that spells could be used in vanilla as a way to jsonize monster special attacks, making them a lot more flexible.
#### Describe the solution

I added a new mattack_actor that creates a spell, and its call() function casts the spell at its target. Currently, this only will work for spells with the attack type spell effect, and summons. This can be expanded at a later date.
#### Describe alternatives you've considered

making a new mattack for every single special attack i want to do for magiclysm.
#### Additional context

monster burning hands
monster fireball

### CODE REVIEW

File: src/magic.cpp [hunk 1]
```diff
@@ -623,11 +623,11 @@ bool spell::is_valid_target( valid_target t ) const
     return type->valid_targets[t];
 }
 
-bool spell::is_valid_target( const tripoint &p ) const
+bool spell::is_valid_target( const Creature &caster, const tripoint &p ) const
 {
     bool valid = false;
     if( Creature *const cr = g->critter_at<Creature>( p ) ) {
-        Creature::Attitude cr_att = cr->attitude_to( g->u );
+        Creature::Attitude cr_att = cr->attitude_to( caster );
```
Comment (BevapDin) [inline src/magic.cpp]: As you have changed it here: `g->u` is still used a few lines down in this very same function. You may want to add a comment if it's intentional that `g->u` is used and not `caster`.

**Comment by ZhilkinSerg** [comment]
Will monsters have mana?

**Comment by KorGgenT** [comment]
no, monsters use the cooldown of special attacks. i think that's perfectly reasonable to use, rather than putting `known_magic` all the way down in `Creature`

**Comment by KorGgenT** [comment]
jenkins rebuild

**Comment by Ignaramico** [comment]
More trauma from the owlbear...

Cant it be something like kind of the robots/turrets weapons?, i mean having limited uses (like the turrets ammo) of the spells, as if they were carrying a wand and using it instead of a cooldown, or reduced range distance and AoE for its spells, making the monster more of hard to attack than outright evading it.

I have encountered 2 owlbears and they are fast to close the distance making ranged attacks of less than 20 tiles kind of worthless add that to impossible to evade spells, and hell ensues.
And thats without taking into account other things like the bulette hiding in the ground, would be worse than a yugg and its 20CD dart

**Comment by KorGgenT** [comment]
owlbears don't actually cast any spells. and no, it's going to remain a cooldown

**Comment by Ignaramico** [comment]
ohhh, i thought it would be all the monsters

**Comment by I-am-Erk** [comment]
Basically this is a way to jsonize adding crazy special effects to monsters.

**Comment by Ignaramico** [comment]
For some reason i thought it would be every magiclysm monster that would be able to cast spells. Still im hoping that the CD time are balanced or very high

**Comment by I-am-Erk** [comment]
Depends on the spell. A monster might get a weak spell as its only attack, or a powerful rare special move

**Comment by KorGgenT** [comment]
"would be able to" is, in this case, misleading. this PR allows content creators to assign spells as special attacks for monsters. as such, monsters will or will not gain a spell attack, based on their json entry.

**Comment by gc2020x4** [comment]
@kevingranade @KorGgenT 
Kevin Grenade must leave the project forever.

**Pull request merged**